Animal filters functionality

Alexandru-Bulai opened the pull request:
# PR 26-01-24
## Overview
In this PR we have the main functionality to select a single type of animal 
## Key Changes :
 1) JS file :
    - Added gallery.js file to code functionality related to the page
    - Implemented core functionality to select a single type of animal

![Jan-26-2024 10-58-41](https://github.com/Alexandru-Bulai/website-project/assets/114404620/aa6f378a-7999-4148-8c98-5ab89cc8bd72)

Alexandru-Bulai committed:
add gallery.js file and script to gallery.html

Alexandru-Bulai committed:
js changes

Alexandru-Bulai committed:
single animal selection filter functionality

Alexandru-Bulai committed:
lintern fix

vladiiancu commented on public/JavaScript/gallery.js:
```diff
@@ -0,0 +1,45 @@
+document.addEventListener('DOMContentLoaded', () => {
+  // Object mapping animal types to their respective classes
+  const animals = {
+    dogs: '.dog',
```
Having logic tied to a CSS class should be avoided when possible. What other options can you think of?

Alexandru-Bulai replied:
i can work with data attributes instead so im not tied to css anymore

Alexandru-Bulai committed:
data attributes filtering method

Alexandru-Bulai committed:
missed description functionality on js file

vladiiancu commented on public/JavaScript/gallery.js:
```diff
@@ -0,0 +1,33 @@
+document.addEventListener('DOMContentLoaded', () => {
```
Good choice since it doesn't wait for images to load

vladiiancu commented on public/JavaScript/gallery.js:
```diff
@@ -0,0 +1,33 @@
+  let selectedAnimalType = null
+
+  // Function to hide all gallery items
+  const hideAllAnimals = () => {
```
This function can receive an argument instead of searching form it inside the function. (Dependency injection)

This way the function doesn't require anything else, can be easily tested etc

vladiiancu reviewed (changes requested):
I would like to have things a bit more independent of one another, generally speaking.

But it's a good first version.

I don't see any tests thought

Alexandru-Bulai committed:
make hideAllAnimals function take arguments

Alexandru-Bulai committed:
lintern fix

vladiiancu commented on public/JavaScript/gallery.js:
```diff
@@ -0,0 +1,33 @@
+  const showSelectedAnimals = () => {
+    hideAllAnimals(document.querySelectorAll('.gallery-items'))
+    if (selectedAnimalType) {
+      document.querySelectorAll(`.gallery-items[data-animal-type="${selectedAnimalType}"]`).forEach(item => {
```
you already looked for and stored all the gallery-items. You don't need to go and search all of them again

vladiiancu commented on public/JavaScript/gallery.js:
```diff
@@ -0,0 +1,33 @@
+  // Target animal type
+  document.querySelectorAll('[data-animal-type]').forEach(filterOption => {
+    filterOption.addEventListener('click', () => {
+      selectedAnimalType = filterOption.dataset.animalType
```
When I click on an item, I change an almost global variable. So an action here, changes the state of the app somewhere else.

These kind of side-effects should be avoided.

Alexandru-Bulai committed:
encapsule <selectedAnimalType>to prevent side effect

Alexandru-Bulai committed:
lintern fix

Alexandru-Bulai committed:
Bug fix

Fixed bug on gallery item

When we were clicking on the gallery image 'dog' we were also selecting it, but the intent in here is to select a specific animal only when we click on the filter not the gallery image

vladiiancu commented on public/JavaScript/gallery.js:
```diff
@@ -8,18 +9,24 @@ document.addEventListener('DOMContentLoaded', () => {
+	};
   // Target animal type
+	function setSelectedAnimalType(type) {
+		selectedAnimalType = type;
```
This is not really encapsulation. You refactor the responsibility of changing the global state of selectedAnimalType to a single function. 
The point is to not use a global state at all. 

So instead of setting the selectedAnimalType and then using it in another function, I can just call the function which uses it and send it as parameter the selection made.

Alexandru-Bulai committed:
remove global function

Moved the global function used to target the data-animal-type intothe function  'setSelectedAnimalType' to avoid side effects by having a global variable

vladiiancu reviewed (approved):
Let's move forward for now and we will revisit this later
